src/crepe/system/AnimatorSystem.cpp

```diff
@@ -13,12 +12,27 @@ void AnimatorSystem::update() {
 
 	RefVector<Animator> animations = mgr.get_components_by_type<Animator>();
 
-	uint64_t tick = SDLContext::get_instance().get_ticks();
+	double elapsed_time = this->timer.get_current_time();
+
 	for (Animator & a : animations) {
 		if (!a.active) continue;
-		// (10 frames per second)
-		a.curr_row = (tick / 100) % a.row;
-		a.spritesheet.mask.x = (a.curr_row * a.spritesheet.mask.w) + a.curr_col;
-		a.spritesheet.mask = a.spritesheet.mask;
+
+		Animator::Data & ctx = a.data;
+		double frame_duration = 1.0f / ctx.fps;
+
+		int last_frame = ctx.curr_row;
+
+		int cycle_end = (ctx.cycle_end == -1) ? ctx.row : ctx.cycle_end;
+		int total_frames = cycle_end - ctx.cycle_start;
+
+		int curr_frame = static_cast<int>(elapsed_time / frame_duration) % total_frames;
+
+		ctx.curr_row = ctx.cycle_start + curr_frame;
+		ctx.spritesheet.mask.x = ctx.curr_row * ctx.spritesheet.mask.w;
+		ctx.spritesheet.mask.y = (ctx.curr_col * ctx.spritesheet.mask.h);
```

Comment:
Dit horen allemaal private variabelen (van de Components met friend AnimatorSystem) te zijn, de user mag hier vooral niet aankomen, toch?

Reply:
dit is in system toch de user komt hier nooit.

Reply:
Dit is inderdaad in een system waar de user nooit komt, helemaal correct. Maar dat bedoel ik niet.
Je slaat hier data op in `Animator::data` (= een **public** struct die **alle** data van Animator bevat). De game programmeur (aka user) kan dus bij deze data komen (want het is public). Maar, volgens mij, is deze data alleen bestemd voor dit specifieke system (als tussentijds opslagmedium) en zou deze data dus private moeten zijn. Onderscheidt maken tussen private/public data kan nu niet omdat je een struct opslaat die alle data bevat, maar dit ga je als het goed is veranderen.

Reply:
de user mag bij alle data van de animator komen en sprite maar de mask kot die niet bij want die is private